### PR TITLE
Create dataset application class

### LINKED ISSUE #393: Implement Dataset application-level object

### PR DESCRIPTION
**Pull request by Bento007**
- Adding dataset application class
- Adding TransformBase to simplify transforming DBobjects
- Adding custom JSON encoder to handle dateTime object return from DB .

closes #393

### CODE REVIEW

File: backend/corpora/common/corpora_orm.py [hunk 1]
```diff
@@ -230,6 +270,10 @@ class DbDataset(Base):
     created_at = Column(DateTime, nullable=False, server_default=DEFAULT_DATETIME)
     updated_at = Column(DateTime, nullable=False, server_default=DEFAULT_DATETIME)
 
+    artifacts = relationship("DbDatasetArtifact", back_populates="dataset")
```
Comment (maniarathi) [inline backend/corpora/common/corpora_orm.py:273]: We probably also want to add the `cascade` argument to inform that children in a relationship should be affected when the parent changes (i.e. gets deleted). Ditto in all the child relationships in this file.

File: backend/corpora/common/corpora_orm.py [hunk 2]
```diff
@@ -14,18 +15,57 @@
...
 from common.corpora_config import CorporaDbConfig
+from common.utils.json import CustomJSONEncoder
 
 
-Base = declarative_base()
-deployment_stage = os.environ["DEPLOYMENT_STAGE"]
+class TransformingBase(object):
+    """
+    Add functionality to transform a Base object, and recursively transform its linked entities.
+
```
Comment (maniarathi) [inline backend/corpora/common/corpora_orm.py:32]:
```suggestion
```

File: backend/corpora/common/corpora_orm.py [hunk 3]
```diff
@@ -14,18 +15,57 @@
...
+
```
Comment (maniarathi) [inline backend/corpora/common/corpora_orm.py:33]:
```suggestion
```

File: backend/corpora/common/entities/dataset.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+from .entity import Entity
+from ..corpora_orm import DbDataset
+
+
+class Dataset(Entity):
+    tables = [
+        DbDataset,
+    ]
+
+    def __init__(self, db_obj: DbDataset):
```
Comment (maniarathi) [inline backend/corpora/common/entities/dataset.py:12]: I'm not sure that the Dataset entity class should be initialized with a DbDataset. The primary use case of this class is to be able to take in the input from a user who is inputting a dataset and parse it into a Dataset object that can be transformed into a DbDataset object to be placed into the database. For this, we actually do need all the attributes available to set similar to the function signature in project.py.

File: backend/corpora/common/corpora_orm.py [hunk 5]
```diff
@@ -14,18 +15,57 @@
...
 pkg_root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))  # noqa
 sys.path.insert(0, pkg_root)  # noqa
 
 from common.corpora_config import CorporaDbConfig
+from common.utils.json import CustomJSONEncoder
 
 
-Base = declarative_base()
-deployment_stage = os.environ["DEPLOYMENT_STAGE"]
+class TransformingBase(object):
```
Comment (maniarathi) [inline backend/corpora/common/corpora_orm.py:29]: I'm curious about this class as a whole. I don't see `to_dict` or `to_json` being used anywhere -- what's the use case you're thinking of here and why specifically convert the database object to dictionary and JSON instead of the application layer objects?

Also what's the benefit of having a class in the ORM layer versus a utility function that could perform the transformation?

Comment (Bento007) [reply backend/corpora/common/corpora_orm.py:29]: The primary use case is when a user make an api request they will want a json response. 

The benefit of having it in the ORM layer is that we only need one conversion function for all of the DB objects. It also allows us to follow the relationships defined by the ORM and convert those object as well in a single function.

File: backend/corpora/common/corpora_orm.py (hunk 1 above)
Comment (Bento007) [reply backend/corpora/common/corpora_orm.py:273]: Good idea, I'll add to all of the children in this class.

File: backend/corpora/common/entities/dataset.py (hunk 4 above)
Comment (Bento007) [reply backend/corpora/common/entities/dataset.py:12]: In that case I think we should create a function call `new` or `create` that takes in all of those parameters, and  generates a DBDataset to pass into the `__init__`. It would be similar to get, except no querying.

File: backend/corpora/common/corpora_orm.py (hunk 1 above)
Comment (Bento007) [reply backend/corpora/common/corpora_orm.py:271]: On second thought I think this should wait until we are handling deletion, since this feature will not be tested until then.

File: backend/corpora/common/entities/dataset.py (hunk 4 above)
Comment (Bento007) [reply backend/corpora/common/entities/dataset.py:12]: These details can be fleshed out when we add save functionality.

File: backend/corpora/common/corpora_orm.py [hunk 6]
```diff
@@ -267,7 +271,7 @@ class DbDeploymentDirectory(Base):
     created_at = Column(DateTime, nullable=False, server_default=DEFAULT_DATETIME)
     updated_at = Column(DateTime, nullable=False, server_default=DEFAULT_DATETIME)
 
-    dataset = relationship("DbDataset")
+    dataset = relationship("DbDataset", back_populates="deployment_directories")
 
 
 class DbContributor(Base):
```
Comment (maniarathi) [inline backend/corpora/common/corpora_orm.py:277]: Missing a relationship here to DbDatasetContributor

File: backend/corpora/common/corpora_orm.py [hunk 7]
```diff
@@ -299,3 +303,6 @@ class DbDatasetContributor(Base):
     dataset_id = Column(ForeignKey("dataset.id"), nullable=False)
     created_at = Column(DateTime, nullable=False, server_default=DEFAULT_DATETIME)
     updated_at = Column(DateTime, nullable=False, server_default=DEFAULT_DATETIME)
+
+    contributor = relationship("DbContributor")
```
Comment (maniarathi) [inline backend/corpora/common/corpora_orm.py:307]: You'll have to back_populate here once the relationship is added to DbContributors.

File: backend/corpora/common/entities/dataset.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        self.db_obj = db_obj
+
+    @classmethod
+    def _query(cls, key: str) -> typing.List["Dataset"]:
```
Comment (maniarathi) [inline backend/corpora/common/entities/dataset.py:16]: Add docstring? I find it good practice to have it even for "self-explanatory" functions because it forces that the implementation matches what we meant to do written in plain English.

File: backend/corpora/common/entities/dataset.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        result = cls.db.query(table_args=[DbDataset], filter_args=[key == DbDataset.id])
+        return result
+
+    @classmethod
+    def _load(cls, db_result: typing.List["Dataset"]) -> "Dataset":
```
Comment (maniarathi) [inline backend/corpora/common/entities/dataset.py:21]: Docstring please?

File: tests/unit/backend/corpora/fixtures/database/__init__.py [hunk 10]
```diff
@@ -98,11 +100,17 @@ def _create_test_datasets(self):
...
+            dataset_id=test_dataset_id,
         )
         self.db.session.add(project_dataset)
         self.db.session.commit()
 
+        deployment_directory = DbDeploymentDirectory(
+            id="test_deployment_directory_id", dataset_id=test_dataset_id, environment="test", url="test_url"
+        )
+        self.db.session.add(deployment_directory)
+        self.db.session.commit()
```
Comment (maniarathi) [inline tests/unit/backend/corpora/fixtures/database/__init__.py:112]: Since you've added a `commit` here, you could delete line 106 so that a single transaction occurs within this single function.

File: backend/corpora/common/corpora_orm.py [hunk 11]
```diff
@@ -230,6 +230,10 @@ class DbDataset(Base):
     created_at = Column(DateTime, nullable=False, server_default=DEFAULT_DATETIME)
     updated_at = Column(DateTime, nullable=False, server_default=DEFAULT_DATETIME)
```
[the remark this answers is not in the document]
Comment (Bento007) [reply backend/corpora/common/corpora_orm.py:232]: added

File: backend/corpora/common/corpora_orm.py (hunk 6 above)
Comment (Bento007) [reply backend/corpora/common/corpora_orm.py:277]: is that relationship needed here

File: tests/unit/backend/corpora/fixtures/database/__init__.py (hunk 10 above)
Comment (Bento007) [reply tests/unit/backend/corpora/fixtures/database/__init__.py:112]: fixing

File: backend/corpora/common/entities/dataset.py (hunk 9 above)
Comment (Bento007) [reply backend/corpora/common/entities/dataset.py:21]: fixed

File: backend/corpora/common/entities/dataset.py (hunk 8 above)
Comment (Bento007) [reply backend/corpora/common/entities/dataset.py:16]: fixed

**Comment by Bento007** [comment]
adding issue to apply cascade to relationships https://github.com/chanzuckerberg/corpora-data-portal/issues/414

File: backend/corpora/common/corpora_orm.py [hunk 12]
```diff
@@ -230,6 +230,10 @@ class DbDataset(Base):
...
 
+    artifacts = relationship("DbDatasetArtifact", back_populates="dataset")
+    deployment_directories = relationship("DbDeploymentDirectory", back_populates="dataset")
+    contributors = relationship("DbDatasetContributor")
```
[the remark this answers is not in the document]
Comment (Bento007) [reply backend/corpora/common/corpora_orm.py]: fix i think

**Comment by Bento007** [comment]
For the issues with project, I will save that for a future PR.

**Comment by Bento007** [comment]
In the sqlAlchemy is suggest using back_populates. If you would like to implement using backref, I wouldn't be offended, but I don't see a compelling reason to make the change at this moment.

File: backend/corpora/common/entities/dataset.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        DbDataset,
+    ]
+
+    def __init__(self, db_obj: DbDataset):
+        self.db_obj = db_obj
+
+    @classmethod
+    def _query(cls, key: str) -> typing.List["Dataset"]:
+        """
+        Queries the database for required entity data
```
Comment (maniarathi) [inline backend/corpora/common/entities/dataset.py:18]:
```suggestion
        Given a key representing a dataset UUID, queries the database to return the associated dataset data.
```

File: backend/corpora/common/entities/dataset.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        :param key: dataset.id is the primary key.
+        :return: list of query result rows
+        """
+        result = cls.db.query(table_args=[DbDataset], filter_args=[key == DbDataset.id])
+        return result
+
+    @classmethod
+    def _load(cls, db_result: typing.List["Dataset"]) -> "Dataset":
+        """
+        Parses a database query result into an Entity instance
```
Comment (maniarathi) [inline backend/corpora/common/entities/dataset.py:28]:
```suggestion
        Parses a database query result into a Dataset Entity instance.
```
